docs/rfcs/0012-introduce-ui-store.md

```diff
@@ -0,0 +1,145 @@
+* Fewer resources needed for context shifting between threads.
+
+For an example of thread simplicity, an `Engine` typically has its own 'engine thread' to perform async work and post/request results to the main thread (these APIs are identified with the `@UiThread` annotation). Once we get the callback for those results, we then need to dispatch an action to the store that will then happen on a `Store` thread. Feature components then observe for state changes and then make UI changes on the main thread. A simplified form of this thread context switching can be seen in the example below:
+
+```kotlin
+// engine thread
+engineView.requestApiResult { result ->
+  // received on the main thread.
+  store.dispatch(UpdateResultAction(result))
+}
+
+// store thread
+fun reduce(state: State, action: Action) {
+  is UpdateResultAction -> {
+    // do things here.
+  }
+}
+
+// store thread
+Middleware {
+  override fun invoke(
+    context: MiddlewareContext<State, Action>,
+    next: (Action) -> Unit,
+    action: Action,
+  ) {
+    // perform side-effects that also happen on the store thread.
+  }
+}
+
+// main thread
+store.flowScoped { flow ->
+  flow.collect {
+    MainScope().launch {
+      // perform work on the main thread.
+    }
+  }
+}
+```
+
+With the changes in this RFC, this switching of threads can be reduced (notable comments marked with 📝):
+
+```kotlin
+// engine thread
+engineView.requestApiResult { result ->
+  // received on the main thread.
+  store.dispatch(UpdateResultAction(result))
+}
+
+// 📝 main thread - now on the same thread, processed immediately.
+fun reduce(state: State, action: Action) {
+  is UpdateResultAction -> {
+    // do things here.
+  }
+}
+
+// 📝 main thread - now on the same thread, processed immediately.
+Middleware {
+  override fun invoke(
+    context: MiddlewareContext<State, Action>,
+    next: (Action) -> Unit,
+    action: Action,
+  ) {
+    // 📝 perform side-effects that now happen on the main thread.
+  }
+}
+
+// main thread
+store.flowScoped { flow ->
+  flow.collect {
+    MainScope().launch { // 📝 no longer necessary to launch a new coroutine unless desired.
+      // perform work on the main thread.
+    }
+  }
+}
+```
+
+## Guide-level explanation
+
+Extending the existing `Store` class to use the `Dispatchers.Main.immediate` will ensure that UI stores will stay on the same UI thread and have that work done immediately. Using a distinct class named `UiStore` also makes it clear to the developer that this is work that will be done on the UI thread and its implications will be made a bit more clear when it's used.
+
+```kotlin
+@MainThread
+open class UiStore<S : State, A : Action>(
+    initialState: S,
+    reducer: Reducer<S, A>,
+    middleware: List<Middleware<S, A>> = emptyList(),
+    threadNamePrefix: String? = null,
+) : Store<S, A>(
+    initialState,
+    reducer,
+    middleware,
+    threadNamePrefix,
+) {
+  override val dispatcher = Dispatchers.Main.immediate
```

Comment:
For this `override` to work, we'd need to `open` the dispatcher on the `Store`, which would allow any `Store` subtype to have complete control over its threading model. If we're going that route, I'd still prefer moving the dispatcher to be an optional constructor parameter to the `Store` instead of using inheritance because I think it makes for cleaner code and easier overrides under test.

For example:
```
// this implementation:
val ioStore = object : Store<State, Action> {
    override val dispatcher = Dispatcher.IO
}

// vs constructor implementation
val ioStore = Store<State, Action>(dispatcher = Dispatcher.IO)
```

Reply:
And we could still add a specific `UiStore` for clarity:

```
open class UiStore<S, A> : Store<S, A>(dispatcher = Dispatcher.Main.immediate)
```

Reply:
Addressed [here](https://github.com/mozilla-mobile/firefox-android/pull/5353#issuecomment-1974174872).